### PR TITLE
[APP-1318] `SelectMediaButton`

### PR DESCRIPTION
**Pull request by estrattonbailey**
Stacked on #8839

Combines the image/video buttons into a single button. Provides the following validations:
- limits selection of images to 4
  - on iOS we can do this in the image picker itself
  - on Android, we do this validation manually
- limits selection of videos and GIFs to 1
  - provides separate errors for video and GIF files
- disallows videos that are longer than 3 minutes
- prevents selecting a mix of media types
  - uses whatever selected item is returned first as the "type" to constrain against
    - example: select 5 images and a video, we'll only take the first 4 images
  - fixes a prod bug that allows you to select an image first, then a GIF (which will be rasterized)

### CODE REVIEW

File: src/view/com/composer/Composer.tsx [hunk 1]
```diff
@@ -514,13 +522,6 @@ export const ComposePost = ({
       onPostSuccess?.(postSuccessData)
     }
     onClose()
-    Toast.show(
-      thread.posts.length > 1
-        ? _(msg`Your posts have been published`)
-        : replyTo
-          ? _(msg`Your reply has been published`)
-          : _(msg`Your post has been published`),
-    )
```
Comment (surfdude29) [inline src/view/com/composer/Composer.tsx]: I think these toasts give a useful confirmation to the user, just wondering why they're being removed? 🤔

File: src/view/com/composer/Composer.tsx [hunk 2]
```diff
@@ -811,7 +816,9 @@ let ComposerPost = React.memo(function ComposerPost({
         if (isNative) return // web only
         const [mimeType] = uri.slice('data:'.length).split(';')
         if (!SUPPORTED_MIME_TYPES.includes(mimeType as SupportedMimeTypes)) {
-          Toast.show(_(msg`Unsupported video type`), 'xmark')
+          toast.show(_(msg`Unsupported video type`), {
```
Comment (surfdude29) [inline src/view/com/composer/Composer.tsx]: The current error toast gives the mime type of the video in the error message, would that be useful to include here?
https://github.com/bluesky-social/social-app/blob/f001af2fe3ca977beb22fefefb2f98ba2c076d49/src/view/com/composer/videos/SelectVideoBtn.tsx#L48

File: src/view/com/composer/SelectMediaButton.tsx [hunk 3]
```diff
@@ -0,0 +1,501 @@
...
+
+    if (typeof selectedVideo.duration !== 'number') {
+      try {
+        const metadata = await getAdditionalVideoMetadata(selectedVideo)
+        selectedVideo.duration = metadata.duration
+        selectedVideo.width = metadata.width
+        selectedVideo.height = metadata.height
+      } catch (e: any) {
+        logger.error(`processSelectedAssets: failed to get video metadata`, {
+          safeMessage: e.message,
```
[the remark this answers is not in the document]
Comment (mozzius) [reply src/view/com/composer/SelectMediaButton.tsx]: ok so replace the type assertion with a different, entirely the same kind of assertion.

File: src/view/com/composer/SelectMediaButton.tsx [hunk 4]
```diff
@@ -0,0 +1,501 @@
...
+  } else if (mimeType?.startsWith('image/')) {
+    type = 'image'
+  }
+
+  // console.log({
+  //   asset,
+  //   type,
+  //   mimeType,
+  // })
+
```
[the remark this answers is not in the document]
Comment (mozzius) [reply src/view/com/composer/SelectMediaButton.tsx]: What

File: src/view/com/composer/SelectMediaButton.tsx [hunk 5]
```diff
@@ -0,0 +1,501 @@
...
+      const errors = Array.from(errorCodes).map(error => {
+        return {
+          [SelectedAssetError.Unsupported]: _(
+            msg`One or more of your selected files are not supported.`,
+          ),
+          [SelectedAssetError.MixedTypes]: _(
+            msg`Selecting multiple media types is not supported.`,
+          ),
+          [SelectedAssetError.MaxImages]: _(
+            msg`You can select up to ${MAX_IMAGES} total images.`,
```
Comment (surfdude29) [inline src/view/com/composer/SelectMediaButton.tsx]:
```suggestion
            msg`You can select up to 4 total images.`,
```

We currently hardcode the maximum number of images in this string (I'm not clear on when it's shown):
https://github.com/bluesky-social/social-app/blob/f001af2fe3ca977beb22fefefb2f98ba2c076d49/src/lib/media/picker.shared.ts#L25
Would it make sense to do that here as well? (And if so, the other string could perhaps be aligned with this one to avoid duplication.)

File: src/view/com/composer/text-input/TextInput.tsx [hunk 6]
```diff
@@ -154,6 +154,7 @@ export const TextInput = forwardRef(function TextInputImpl(
       }
 
       const uris = files.map(f => f.uri)
+      // TODO
```
[the remark this answers is not in the document]
Comment (estrattonbailey) [reply src/view/com/composer/text-input/TextInput.tsx]: Oh thanks copilot, mean to rm this

File: src/view/com/composer/SelectMediaButton.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (estrattonbailey) [reply src/view/com/composer/SelectMediaButton.tsx]: nah

File: src/view/com/composer/SelectMediaButton.tsx [hunk 7]
```diff
@@ -0,0 +1,501 @@
...
+    }
+  } else if (selectableAssetType === 'gif') {
+    if (supportedAssets.length > 1) {
+      errors.add(SelectedAssetError.MaxGIFs)
+      supportedAssets = supportedAssets.slice(0, 1)
+    }
+  }
+
+  return {
+    type: selectableAssetType!, // set above
```
[the remark this answers is not in the document]
Comment (estrattonbailey) [reply src/view/com/composer/SelectMediaButton.tsx:353]: validated above, as comment states

File: src/view/com/composer/SelectMediaButton.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (estrattonbailey) [reply src/view/com/composer/SelectMediaButton.tsx]: yep, debug code, will remove

[the remark this answers is not in the document]
Comment (mozzius) [reply src/view/com/composer/SelectMediaButton.tsx]: Oh I think the app isn't showing me the diff properly

File: src/view/com/composer/SelectMediaButton.tsx [hunk 8]
```diff
@@ -0,0 +1,495 @@
...
+type ValidatedImagePickerAsset = Omit<ImagePickerAsset, 'mimeType'> & {
+  mimeType: string
+}
+
+/**
+ * Codes for known validation states
+ */
+enum SelectedAssetError {
+  Unsupported = 'Unsupported',
+  MixedTypes = 'MixedTypes',
```
Comment (APiligrim) [inline src/view/com/composer/SelectMediaButton.tsx:56]: this is a good idea! Makes it a lot easier to track / add on new edge cases

**Comment by mozzius** [comment]
@github-actions ota

File: src/view/com/composer/SelectMediaButton.tsx [hunk 9]
```diff
@@ -0,0 +1,495 @@
...
+  gif: 'image/gif',
+  jpg: 'image/jpeg',
+  jpeg: 'image/jpeg',
+  png: 'image/png',
+  svg: 'image/svg+xml',
+  heic: 'image/heic',
+}
+
+/**
+ * Attemps to bucket the given asset into one of our known types based on its
```
Comment (APiligrim) [inline src/view/com/composer/SelectMediaButton.tsx]: thank you for these comments! They are super helpful!

File: src/view/com/composer/SelectMediaButton.tsx [hunk 10]
```diff
@@ -0,0 +1,495 @@
...
+    sheetWrapper,
+    processSelectedAssets,
+    selectionCountRemaining,
+  ])
+
+  return (
+    <Button
+      testID="openMediaBtn"
+      onPress={onPressSelectMedia}
+      label={_(msg`Media`)}
```
Comment (surfdude29) [inline src/view/com/composer/SelectMediaButton.tsx]:
```suggestion
      label={_(
        msg({
          message: `Add media to post`,
          comment: `Accessibility label for button in composer to add photos or a video to a post`,
        }),
      )}
```
Would a slightly longer label for the button be helpful here? (I'm genuinely not sure how much info it's good to include for screen reader users in a button label like this.)

Also, a comment for translators might be helpful (especially if the label is just `Media`, but perhaps also if it's more descriptive.)

File: src/view/com/composer/SelectMediaButton.tsx [hunk 11]
```diff
@@ -0,0 +1,495 @@
...
+        errors: errorCodes,
+      } = await processImagePickerAssets(rawAssets, {
+        selectionCountRemaining,
+        allowedAssetTypes,
+      })
+
+      /*
+       * Convert error codes to user-friendly messages.
+       */
+      const errors = Array.from(errorCodes).map(error => {
```
Comment (APiligrim) [inline src/view/com/composer/SelectMediaButton.tsx:387]: would these errors show Toast messages to the user to say why their files failed to upload? 👀

**Review by APiligrim — COMMENTED** [review]
the written types for edge cases and separated functions for each of the processes look great and make the code readable too. Thank you for all the detailed explanations!

File: src/view/com/composer/SelectMediaButton.tsx [hunk 12]
```diff
@@ -0,0 +1,495 @@
...
+  selectedAssetsCount,
+  onSelectAssets,
+}: SelectMediaButtonProps) {
+  const {_} = useLingui()
+  const {requestPhotoAccessIfNeeded} = usePhotoLibraryPermission()
+  const {requestVideoAccessIfNeeded} = useVideoLibraryPermission()
+  const sheetWrapper = useSheetWrapper()
+  const t = useTheme()
+
+  const selectionCountRemaining = MAX_IMAGES - selectedAssetsCount
```
Comment (APiligrim) [inline src/view/com/composer/SelectMediaButton.tsx:371]: I like that MAX_IMAGES is a variable in cases later this variable gets updated, we only need to update it once

File: src/view/com/composer/SelectMediaButton.tsx [hunk 13]
```diff
@@ -0,0 +1,495 @@
...
+ * Codes for known validation states
+ */
+enum SelectedAssetError {
+  Unsupported = 'Unsupported',
+  MixedTypes = 'MixedTypes',
+  MaxImages = 'MaxImages',
+  MaxVideos = 'MaxVideos',
+  VideoTooLong = 'VideoTooLong',
+  MaxGIFs = 'MaxGIFs',
+  NoGifsOnNative = 'NoGifsOnNative',
```
Comment (APiligrim) [inline src/view/com/composer/SelectMediaButton.tsx]: for `NoGifsOnNative` are we not going to support GIFs for now at all for native? Just wondering why 👀

Comment (APiligrim) [reply src/view/com/composer/SelectMediaButton.tsx]: Also we are still keeping the GIF button on native then right?

File: src/view/com/composer/Composer.tsx (hunk 1 above)
Comment (estrattonbailey) [reply src/view/com/composer/Composer.tsx]: Ooof, thanks, noticed this in #8710 and totally forgot to add back here, thanks!

File: src/view/com/composer/SelectMediaButton.tsx (hunk 9 above)
Comment (surfdude29) [inline src/view/com/composer/SelectMediaButton.tsx]:
```suggestion
 * Attempts to bucket the given asset into one of our known types based on its
```
micronit

File: src/view/com/composer/Composer.tsx (hunk 2 above)
Comment (estrattonbailey) [reply src/view/com/composer/Composer.tsx]: Ah good idea!

File: src/view/com/composer/SelectMediaButton.tsx (hunk 13 above)
Comment (estrattonbailey) [reply src/view/com/composer/SelectMediaButton.tsx]: Good question! The GIF button is actually for selecting a GIF from Tenor. On web, we allow you to upload a `.gif` file, which we then convert to a video before posting. We haven't implemented this logic yet on native, that's what this error is for.

File: src/view/com/composer/SelectMediaButton.tsx (hunk 11 above)
Comment (estrattonbailey) [reply src/view/com/composer/SelectMediaButton.tsx:387]: Uploading happens in a later step, these errors only pertain to the selection of assets.

File: src/view/com/composer/SelectMediaButton.tsx (hunk 13 above)
Comment (mozzius) [reply src/view/com/composer/SelectMediaButton.tsx]: Is that an error though? Currently they're just handled as images (because they are also valid images)

File: src/view/com/composer/SelectMediaButton.tsx (hunk 5 above)
Comment (estrattonbailey) [reply src/view/com/composer/SelectMediaButton.tsx]: Oh ya know what, this is just old sloppy logic that we don't need anymore, thanks for mentioning this, I'm gonna just clean this up 👍

File: src/view/com/composer/SelectMediaButton.tsx (hunk 10 above)
Comment (estrattonbailey) [reply src/view/com/composer/SelectMediaButton.tsx]: Yes that's perfect, thanks!

File: src/view/com/composer/SelectMediaButton.tsx (hunk 13 above)
Comment (estrattonbailey) [reply src/view/com/composer/SelectMediaButton.tsx]: Oh this is true, maybe this is too heavy handed. No reason we couldn't also toast the user to tell them it will upload as a static image 🤔

File: src/view/com/composer/SelectMediaButton.tsx [hunk 14]
```diff
@@ -0,0 +1,506 @@
...
+      const errors = Array.from(errorCodes).map(error => {
+        return {
+          [SelectedAssetError.Unsupported]: _(
+            msg`One or more of your selected files are not supported.`,
+          ),
+          [SelectedAssetError.MixedTypes]: _(
+            msg`Selecting multiple media types is not supported.`,
+          ),
+          [SelectedAssetError.MaxImages]: _(
+            msg`You can select up to ${MAX_IMAGES} total images.`,
```
Comment (surfdude29) [inline src/view/com/composer/SelectMediaButton.tsx]:
```suggestion
          [SelectedAssetError.MaxImages]: _(
            msg({
              message: `You can select up to ${plural(MAX_IMAGES, {
                other: '# images',
              })} in total.`,
              comment: `Error message for maximum number of images that can be selected to add to a post, currently 4 but may change.`,
            }),
          ),
```
Would it be possible to add plural formatting and a comment for translators here please, for languages with complex plurals? I also suggested a slight wording tweak to `… ${MAX_IMAGES} images in total.`

File: src/view/com/composer/SelectMediaButton.tsx [hunk 15]
```diff
@@ -0,0 +1,506 @@
...
+      )}
+      accessibilityHint={
+        isNative
+          ? _(
+              msg`Opens device gallery to select up to ${MAX_IMAGES} images, or a single video.`,
+            )
+          : _(
+              msg`Opens device gallery to select up to ${MAX_IMAGES} images, or a single video or GIF.`,
+            )
+      }
```
Comment (surfdude29) [inline src/view/com/composer/SelectMediaButton.tsx]:
```suggestion
      accessibilityHint={
        isNative
          ? _(
              msg({
                message: `Opens device gallery to select up to ${plural(
                  MAX_IMAGES,
                  {
                    other: '# images',
                  },
                )}, or a single video.`,
                comment: `Accessibility hint on native for button in composer to add images or a video to a post. Maximum number of images that can be selected is currently 4 but may change.`,
              }),
            )
          : _(
              msg({
                message: `Opens device gallery to select up to ${plural(
                  MAX_IMAGES,
                  {
                    other: '# images',
                  },
                )}, or a single video or GIF.`,
                comment: `Accessibility hint on web for button in composer to add images, a video, or a GIF to a post. Maximum number of images that can be selected is currently 4 but may change.`,
              }),
            )
      }
```
And if plural formatting and comments could be added here too that would be great 🙏

File: src/view/com/composer/SelectMediaButton.tsx (hunk 13 above)
Comment (mozzius) [reply src/view/com/composer/SelectMediaButton.tsx]: Decided to just support gifs, see below

**Comment by mozzius** [comment]
Regarding `NoGifsOnNative` - screw it, let's just support gifs on native

https://github.com/user-attachments/assets/5ea732a5-c50f-42f1-be6d-f996d92e3c6c

Cavaets:

- it has to skip the compression step, because the library does not support compressing gifs. however, I think most gifs are gonna be below 100mb (!!) so that's fine. we can just let the server reject them. basically works like web does.
- had to implement a custom preview (also fine)

File: src/view/com/composer/Composer.tsx [hunk 16]
```diff
@@ -650,84 +659,88 @@ export const ComposePost = ({
...
-                  isPartOfThread={thread.posts.length > 1}
-                  isReply={index > 0 || !!replyTo}
-                  isActive={post.id === activePost.id}
-                  canRemovePost={thread.posts.length > 1}
-                  canRemoveQuote={index > 0 || !initQuote}
-                  onSelectVideo={selectVideo}
-                  onClearVideo={clearVideo}
-                  onPublish={onComposerPostPublish}
-                  onError={setError}
+      <RootSiblingParent>
```
Comment (mozzius) [inline src/view/com/composer/Composer.tsx:662]: Is this still necessary with Sonner toasts?

File: src/view/com/composer/SelectMediaButton.tsx [hunk 17]
```diff
@@ -0,0 +1,500 @@
...
+      style={a.p_sm}
+      variant="ghost"
+      shape="round"
+      color="primary"
+      disabled={disabled}>
+      <ImageIcon
+        size="lg"
+        style={disabled && t.atoms.text_contrast_low}
+        accessibilityIgnoresInvertColors={true}
+      />
```
Comment (mozzius) [inline src/view/com/composer/SelectMediaButton.tsx:521]: Do we want a new icon here?

**Review by mozzius — CHANGES_REQUESTED** [review]
Looking pretty solid! Only one noteworthy bug - can't seem to select additional images after selecting some

  

https://github.com/user-attachments/assets/3aef42f1-264b-45d2-9fb1-10d3d017c1fb

**Comment by surfdude29** [comment]
I did some testing on the preview branch and it's working well, very nice work @APiligrim, @estrattonbailey and @mozzius  👌

One thing I noticed is that too large videos on web are treated differently – not sure about native, I know everything over 25 MB is compressed there anyway, so maybe it's not an issue – and I think you might want to look again at that.

In prod, when you select a video bigger than 100 MB, you get a specific error message that tells you exactly what the problem is (I think [from here](https://github.com/bluesky-social/social-app/blob/221623f55aa6c1bbe699c8d409832da110923c76/src/view/com/composer/state/video.ts#L395)? because the other version looks like an upload error).

However, when you select a video bigger than 100 MB on the preview branch, you only get the generic `Unsupported` error toast. Tbh this feels quite a lot less useful in understanding why the video can't be added to the post, so if it's possible to show the `VideoTooLargeError` string instead, I think that would be ideal.

| error shown in prod | error shown on preview branch |
|--------|--------|
| IMG_8093 | IMG_8094 |

___

Also, on web, now that some video errors are presented as toasts popping up in the lower-left corner of the screen rather than being shown directly in the composer, I think it might be worth considering using the longer, 6-second toast, to give users more of a chance to spot it and read it.

File: src/view/com/composer/SelectMediaButton.tsx (hunk 17 above)
Comment (estrattonbailey) [reply src/view/com/composer/SelectMediaButton.tsx:521]: I think the image is fine, did you have an idea?

File: src/view/com/composer/Composer.tsx (hunk 16 above)
Comment (estrattonbailey) [reply src/view/com/composer/Composer.tsx:662]: Ah no, thanks

File: src/view/com/composer/SelectMediaButton.tsx (hunk 17 above)
Comment (mozzius) [reply src/view/com/composer/SelectMediaButton.tsx:521]: Looking at our options, there's this one? I think I'm fine with photo but I don't know if it'll be confusing for users if it looks like the video button has disappeared

Comment (mozzius) [reply src/view/com/composer/SelectMediaButton.tsx:521]: There's also the Cat Videos icon lmao

Screenshot 2025-08-14 at 18 07 11

File: src/view/com/composer/SelectMediaButton.tsx [hunk 18]
```diff
@@ -0,0 +1,522 @@
...
+          ),
+          [SelectedAssetError.MaxImages]: _(
+            msg({
+              message: `You can select up to ${plural(MAX_IMAGES, {
+                other: '# images',
+              })} in total.`,
+              comment: `Error message for maximum number of images that can be selected to add to a post, currently 4 but may change.`,
+            }),
+          ),
+          ),
```
Comment (surfdude29) [inline src/view/com/composer/SelectMediaButton.tsx]:
```suggestion
```
Oops, I added a duplicate line at the end here in my suggestion, sorry 😬

Comment (estrattonbailey) [reply src/view/com/composer/SelectMediaButton.tsx]: Np at all!

**Comment by estrattonbailey** [comment]
@surfdude29 when you were testing with a large video on web, what file format was it? If I try to upload a 150mb `.mp4` I get the correct error. So I'm wondering if you tried to upload a video that this branch considers unsupported, but that prod supports.

**Comment by surfdude29** [comment]
@estrattonbailey Ah interesting, I'll do a bit more testing and report back if I can reproduce, and if so with what formats 👍

**Comment by surfdude29** [comment]
@estrattonbailey OK sorry for the delay in getting back to you, I tried several times using Big Buck Bunny as the test file ([from here](https://download.blender.org/peach/bigbuckbunny_movies/)).

I should mention btw that I'm testing in Safari on iPadOS 18.6, so it's possible there's some iOS weirdness going on that makes what I'm seeing an edge case?

Anyway, using the raw unmodified file (`big_buck_bunny_1080p_h264.mov`):
9 minutes 56 seconds long, 725 MB

**On preview branch:** gets `Unsupported` error
**In prod:** gets `Video must be less than 3 minutes long` error

I then took the same file and trimmed it in iPadOS Photos app (no re-encoding):
1 minute 13 seconds long, 110 MB ([can be downloaded here](https://transfer.it/t/6KlPX84hGBRB)):

**On preview branch:** gets `Unsupported` error
**In prod:** gets `VideoTooLarge` error

And I see the same pattern with other video formats (e.g. HEVC), so maybe it's an iOS (or Expo?) thing if it works fine on desktop?

**Comment by surfdude29** [comment]
Excellent, it now triggers the new `FileTooBig` error for me every time I select a video larger than 100 MB 🙏

IMG_8337

Also, this isn't a criticism, and I imagine that the scenario where a user tries to upload a GIF larger than 100 MB will be a rare edge case, but I was interested to see that when I try selecting a gargantuan 180 MB GIF that I made to test with, it triggers the `VideoTooLargeError` rather than the `FileTooBig` error. I guess even though an uploaded GIF is (I think) mostly treated as a video, it goes down a different code path that ends up at a different error?

File: src/view/com/composer/SelectMediaButton.tsx [hunk 19]
```diff
@@ -0,0 +1,609 @@
...
+            msg`You can only select one video at a time.`,
+          ),
+          [SelectedAssetError.VideoTooLong]: _(
+            msg`Videos must be less than 3 minutes long.`,
+          ),
+          [SelectedAssetError.MaxGIFs]: _(
+            msg`You can only select one GIF at a time.`,
+          ),
+          [SelectedAssetError.FileTooBig]: _(
+            msg`This file is too large. Maximum size is 100mb.`,
```
Comment (surfdude29) [inline src/view/com/composer/SelectMediaButton.tsx:494]:
```suggestion
            msg`This file is too large. Maximum size is 100 MB.`,
```
nit

**Comment by estrattonbailey** [comment]
@surfdude29 I think my latest commit should fix the GIF error, since it checks the size as an early step.

**Comment by surfdude29** [comment]
@estrattonbailey It does indeed fix the GIF error for me, and it's so much speedier in general at showing the `FileTooBig` error message, especially for very large videos. Really nice work! 💫

File: yarn.lock [hunk 20]
```diff
@@ -11295,7 +11295,7 @@ expo-image-manipulator@~13.1.7:
   dependencies:
     expo-image-loader "~5.1.0"
 
-expo-image-picker@~16.1.4:
+expo-image-picker@^16.1.4:
   version "16.1.4"
```
Comment (mozzius) [inline yarn.lock]: Doesn't look like it actually got bumped? lockfile still points to v16

Comment (estrattonbailey) [reply yarn.lock]: Heh whoops, fixed now

**Comment by surfdude29** [comment]
> … it's so much speedier in general at showing the `FileTooBig` error message, especially for very large videos …

And with the new version of `expo-image-picker` it shows up instantly now, wow! 🚀

File: src/view/com/composer/SelectMediaButton.tsx [hunk 21]
```diff
@@ -0,0 +1,514 @@
...
+            msg`You can only select one video at a time.`,
+          ),
+          [SelectedAssetError.VideoTooLong]: _(
+            msg`Videos must be less than 3 minutes long.`,
+          ),
+          [SelectedAssetError.MaxGIFs]: _(
+            msg`You can only select one GIF at a time.`,
+          ),
+          [SelectedAssetError.FileTooBig]: _(
+            msg`One or more of your selected files is too large. Maximum size is 100MB.`,
```
Comment (surfdude29) [inline src/view/com/composer/SelectMediaButton.tsx]:
```suggestion
            msg`One or more of your selected files is too large. Maximum size is 100 MB.`,
```
nit: maybe with a no-break space between value and unit?

**Comment by estrattonbailey** [comment]
Let's give this a go
